lib/rules/jsx-no-bind.js

```diff
@@ -40,33 +45,97 @@ module.exports = {
+    // bind expression or an arrow function in different block statements
+    const blockVariableNameSets = {};
+
+    function setBlockVariableNameSet(blockStart) {
+      blockVariableNameSets[blockStart] = {
+        arrowFunc: new Set(),
+        bindCall: new Set(),
+        bindExpression: new Set()
+      };
+    }
+
+    function getVariableViolationType(rightNode) {
+      const nodeType = rightNode.type;
+
+      if (
+        !configuration.allowBind &&
+        nodeType === 'CallExpression' &&
+        rightNode.callee.type === 'MemberExpression' &&
+        rightNode.callee.property.type === 'Identifier' &&
+        rightNode.callee.property.name === 'bind'
+      ) {
+        return 'bindCall';
+      } else if (
+        !configuration.allowArrowFunctions &&
+        nodeType === 'ArrowFunctionExpression'
+      ) {
+        return 'arrowFunc';
+      } else if (
+        !configuration.allowBind &&
+        nodeType === 'BindExpression'
+      ) {
+        return 'bindExpression';
+      }
+
+      return null;
+    }
+
+    function addVariableNameToSet(violationType, variableName, blockStart) {
+      blockVariableNameSets[blockStart][violationType].add(variableName);
+    }
+
+    function getBlockStatementAncestors(node) {
+      return context.getAncestors(node).reverse().filter(
+        ancestor => ancestor.type === 'BlockStatement'
+      );
+    }
+
+    function reportVariableViolation(node, name, blockStart) {
+      const blockSets = blockVariableNameSets[blockStart];
+
+      if (blockSets.bindCall.has(name)) {
+        context.report({node: node, message: bindViolationMessage});
+        return true;
+      } else if (blockSets.arrowFunc.has(name)) {
+        context.report({node: node, message: arrowViolationMessage});
+        return true;
+      } else if (blockSets.bindExpression.has(name)) {
+        context.report({node: node, message: bindExpressionViolationMessage});
+        return true;
+      }
+
+      return false;
+    }
+
+    function findVariableViolation(node, name) {
+      getBlockStatementAncestors(node).find(
+        block => reportVariableViolation(node, name, block.start)
+      );
+    }
+
     return {
-      CallExpression: function(node) {
-        const callee = node.callee;
+      BlockStatement(node) {
+        setBlockVariableNameSet(node.start);
+      },
+
+      VariableDeclarator(node) {
+        const blockAncestors = getBlockStatementAncestors(node);
+        const variableViolationType = getVariableViolationType(node.init);
+
         if (
-          !configuration.allowBind &&
-          (callee.type !== 'MemberExpression' || callee.property.name !== 'bind')
+          blockAncestors.length > 0 &&
+          variableViolationType &&
+          node.parent.kind === 'const' // only support const right now
         ) {
-          return;
-        }
-        const ancestors = context.getAncestors(callee).reverse();
```

Comment:
This block of code reports any violation in a component's `render` method. This rule is supposed to only report violations in JSX props, so this is not needed.